models/interactions.sql

```diff
@@ -0,0 +1,195 @@
+{{ 
+  config(
+    materialized='table',
+    unique_key = 'event_id',
+    sort = 'start_tstamp',
+    dist = 'event_id'
+  )
+}}
+
+with prep as (
+
+ 	select
+    e.event_id,
+    e.page_view_id,
+    e.domain_sessionid,
+    e.domain_userid,
+    coalesce(y.player_id, me.html_id, r.content_title) as media_id, -- replace content_title with id field when available
+    round(mp.duration) as duration,
+    coalesce(y.player_id, me.html_id, r.content_title) as title, -- replace ids with title fields when available
+    case when me.media_type = 'audio' then 'audio' else 'video' end as media_type,
+    coalesce(y.schema_vendor||'-'||y.schema_name, me.schema_vendor||'-'||me.schema_name, r.schema_vendor||'-'||r.schema_name)  as media_player_type,
+    e.page_referrer,
+    coalesce(e.page_url, r.content_url) as content_url,
+    e.geo_region_name,
+    e.br_name,
+    e.dvce_type,
+    e.os_name,
+    e.os_timezone,
+    mpe.type as event_type,
+    e.derived_tstamp as start_tstamp, 
+    mp.current_time as media_start_pos,
+    isnull(mp.playback_rate, 1) as playback_rate,
+    coalesce(y.playback_quality, ve.video_height||'x'||ve.video_width, rd.display_height||'x'||rd.display_width) as playback_quality,
+    case when event_type = 'ended' then 100 else mp.percent_progress end percent_progress,
+    mp.muted
+ 		
+		from {{ ref("mp_events") }} as e
+
+    left join {{ source('atomic', 'com_snowplowanalytics_snowplow_media_player_event_1') }} as mpe
+    on mpe.root_id = e.event_id
+
+    left join {{ source('atomic', 'com_snowplowanalytics_snowplow_media_player_1') }} as mp
+    on mp.root_id = e.event_id
+
+    left join {{ source('atomic', 'com_youtube_youtube_1') }} as y
+    on  y.root_id = e.event_id
+
+    left join {{ source('atomic', 'org_whatwg_media_element_1') }} as me
+    on me.root_id = e.event_id
+
+    left join {{ source('atomic', 'org_whatwg_video_element_1') }} as ve
+    on ve.root_id = e.event_id
+
+    left join {{ source('atomic', 'com_roku_video_1') }} as r
+    on r.root_id = e.event_id
+
+    left join {{ source('atomic', 'com_roku_device_info_1') }} as rd
+	  on  rd.root_id = e.event_id
+
+)
+
+, interaction_ends as (
+
+	select
+		page_view_id,
+		media_id,
+		event_id,
+		event_type,
+		start_tstamp,
+		media_start_pos,
+		lead(start_tstamp, 1) over(partition by page_view_id, media_id order by start_tstamp) end_tstamp,
```

Comment:
Thanks for your explanation, it makes sense. In the reworked model it will be added to `interactions` already.

Reply:
Sorry I guess my point here was that it might be cleaner to have the surrogate key defined earlier within the interactions model so it is more obvious what the partition actually represents.

Reply:
I wanted to do that of course but the issue I faced was {{ dbt_utils.surrogate_key(['p.page_view_id', coalesce(y.player_id, me.html_id) ]) }} play_id doesn't work. Do you have a workaround for this perhaps? The only thing I can think of is to add another CTE after prep and before the partitioning CTE just for the surrogate key calculation.

Reply:
I think you just need to wrap the coalesce in quotes i.e. `{{ dbt_utils.surrogate_key(['p.page_view_id', 'coalesce(y.player_id, me.html_id)' ]) }} play_id`.